### PR TITLE
Cleanup the crop code, fix uncapped drop rates, and correct growth rate

### PR DESCRIPTION
**Pull request by C0bra5**
- Provides some cleanup to the crop code
  - Relies on the maxSize function a lot more instead of hard coding it all the time.
  - General code cleanup to help with readability
- Fixes uncapped drop rates, which could occur from multiple sources, but mostly magic metal berries, ardite berries and cobalt berries.
  - Added some guard rails to help with that
  - Consider buffing the drop rates of magic metal, ardite, and cobalt berries since they only drop 1 nugget, which will be very noticeable for those who use those crops. Consider buffing it to 6 to match the TiC ore berries with their necessary growth block under them.
  - Speaking of those, consider removing the block check in the getGain function since the crop can't grow to maturity without a block under it.
  - Closes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/7517
- Correctly applies multipliers in the weightInfluences functions to reflect the specified needs.
  - Crops that had no documented requirements were left alone.
  - The function determines the available "growth points" based on the crop stick's environment variables.
  - By multiplying a variable, you multiply the amount of "growth points" the variable provides.
  - When a seed mentioned that it had heightened or lowered requirements, it actually always did the opposite.
  - Prim Perl crop can actually die now if you try to grow it without good nutrient, hydration and air quality values

### CODE REVIEW

File: src/main/java/com/github/bartimaeusnek/cropspp/crops/TF/KnighmetalCrop.java [hunk 1]
```diff
@@ -42,14 +37,20 @@ protected String hasBlock() {
...
     }
 
     @Override
     public ItemStack getGain(ICropTile crop) {
-        if (crop.getSize() == 4
-                && (crop.isBlockBelow("blockKnightmetal") || !OreDictionary.doesOreNameExist("blockKnightmetal"))) {
-            return new ItemStack(twilightforest.item.TFItems.armorShard).splitStack(4);
-        } else return new ItemStack(twilightforest.item.TFItems.armorShard);
+        if (crop.getSize() < this.maxSize()
+                || !((crop.isBlockBelow(this.hasBlock()) || !OreDictionary.doesOreNameExist(this.hasBlock())))) {
```
Comment (miozune) [inline src/main/java/com/github/bartimaeusnek/cropspp/crops/TF/KnighmetalCrop.java:46]: This is less readable imo, double negation

File: src/main/java/com/github/bartimaeusnek/cropspp/crops/cpp/CobaltBerryCrop.java [hunk 2]
```diff
@@ -15,38 +15,35 @@ public CobaltBerryCrop() {
...
         if (ConfigValues.debug) return 1;
-        return crop.getSize() >= 2 ? 3000 : 500;
+        return crop.getSize() >= this.maxSize() - 2 ? 3000 : 500;
     }
 
     @Override
-    public String[] attributes() {
-        return new String[] { "OreBerry", "Cobalt", "Metal", "Blue" };
+    public ItemStack getGain(ICropTile crop) {
+        return CCropUtility.getCopiedOreStack("nuggetArdite");
```
Comment (miozune) [inline src/main/java/com/github/bartimaeusnek/cropspp/crops/cpp/CobaltBerryCrop.java:46]: bad copy-paste

File: src/main/java/com/github/bartimaeusnek/cropspp/crops/gregtechCrops/StonelillyCrop.java [hunk 3]
```diff
@@ -55,95 +55,67 @@ public String name() {
...
+                return new String[] { color, "Stone", "Shiny" };
+            case "Gray":
+                return new String[] { color, "Stone", "Metal" };
+            case "Yellow":
+                return new String[] { color, "Stone", "Alien" };
+            case "Nether":
+                return new String[] { color, "Stone", "Evil" };
+            default:
+                // if this line executes consider it a UB
+                return null;
```
Comment (miozune) [inline src/main/java/com/github/bartimaeusnek/cropspp/crops/gregtechCrops/StonelillyCrop.java:117]: Probably return non-null for safety

**Comment by miozune** [comment]
Deprecated classes probably can be removed straightforwardly

File: src/main/java/com/github/bartimaeusnek/cropspp/abstracts/BasicBerryCrop.java [hunk 4]
```diff
@@ -51,14 +46,12 @@ public List<String> getCropInformation() {
...
-        return (int) ((double) humidity * 1.2 + (double) nutrients * 0.9 + (double) air * 0.9);
+        return (int) ((double) humidity / 1.2D + (double) nutrients / 0.9D + (double) air / 0.9D);
     }
 
     @Override
     public int growthDuration(ICropTile crop) {
-        if (ConfigValues.debug) return 1;
-        if (crop.getSize() == 2) return 200;
-        return 700;
+        return ConfigValues.debug ? 1 : (crop.getSize() <= this.maxSize() - 1 ? 200 : 700);
```
[the remark this answers is not in the document]
Comment (C0bra5) [reply src/main/java/com/github/bartimaeusnek/cropspp/abstracts/BasicBerryCrop.java]: should have been addressed by https://github.com/GTNewHorizons/Crops-plus-plus/pull/57/commits/38c9838a209ddeed10cd00bfea8786ac8a2a3baf

File: src/main/java/com/github/bartimaeusnek/cropspp/abstracts/BasicNetherBerryCrop.java [hunk 5]
```diff
@@ -51,12 +45,11 @@ public List<String> getCropInformation() {
...
         // Requires no humidity but nutrients.
-        return (int) ((double) humidity * 0.5 + (double) nutrients * 1.5 + (double) air * 1);
+        return (int) ((double) humidity / 0.5D + (double) nutrients / 1.5D + (double) air);
     }
 
     @Override
     public int growthDuration(ICropTile crop) {
-        if (ConfigValues.debug) return 1;
-        return crop.getSize() == 2 ? 300 : 700;
+        return ConfigValues.debug ? 1 : crop.getSize() >= this.maxSize() - 1 ? 300 : 700;
```
[the remark this answers is not in the document]
Comment (C0bra5) [reply src/main/java/com/github/bartimaeusnek/cropspp/abstracts/BasicNetherBerryCrop.java]: should have been addressed by https://github.com/GTNewHorizons/Crops-plus-plus/pull/57/commits/38c9838a209ddeed10cd00bfea8786ac8a2a3baf

File: src/main/java/com/github/bartimaeusnek/cropspp/abstracts/BasicThaumcraftCrop.java [hunk 6]
```diff
@@ -17,8 +17,7 @@ public int tier() {
 
     @Override
     public int growthDuration(ICropTile crop) {
-        if (ConfigValues.debug) return 1;
-        return crop.getSize() == 2 ? 2200 : 1800;
+        return ConfigValues.debug ? 1 : (crop.getSize() == this.maxSize() - 1 ? 2200 : 1800);
```
[the remark this answers is not in the document]
Comment (C0bra5) [reply src/main/java/com/github/bartimaeusnek/cropspp/abstracts/BasicThaumcraftCrop.java]: should have been addressed by https://github.com/GTNewHorizons/Crops-plus-plus/pull/57/commits/38c9838a209ddeed10cd00bfea8786ac8a2a3baf

File: src/main/java/com/github/bartimaeusnek/cropspp/crops/TF/KnighmetalCrop.java (hunk 1 above)
Comment (C0bra5) [reply src/main/java/com/github/bartimaeusnek/cropspp/crops/TF/KnighmetalCrop.java]: should have been addressed by https://github.com/GTNewHorizons/Crops-plus-plus/pull/57/commits/38c9838a209ddeed10cd00bfea8786ac8a2a3baf

File: src/main/java/com/github/bartimaeusnek/cropspp/crops/cpp/CobaltBerryCrop.java (hunk 2 above)
Comment (C0bra5) [reply src/main/java/com/github/bartimaeusnek/cropspp/crops/cpp/CobaltBerryCrop.java]: should have been addressed by https://github.com/GTNewHorizons/Crops-plus-plus/pull/57/commits/38c9838a209ddeed10cd00bfea8786ac8a2a3baf

File: src/main/java/com/github/bartimaeusnek/cropspp/crops/gregtechCrops/StonelillyCrop.java (hunk 3 above)
Comment (C0bra5) [reply src/main/java/com/github/bartimaeusnek/cropspp/crops/gregtechCrops/StonelillyCrop.java]: should have been addressed by https://github.com/GTNewHorizons/Crops-plus-plus/pull/57/commits/38c9838a209ddeed10cd00bfea8786ac8a2a3baf

**Comment by C0bra5** [comment]
All requested changes should have been made

File: src/main/java/com/github/bartimaeusnek/cropspp/crops/gregtechCrops/GarnydniaCrop.java [hunk 7]
```diff
@@ -60,8 +65,9 @@ public int stat(int n) {
     }
 
     @Override
-    public String[] attributes() {
-        return new String[] { "Shiny", "Crystal", "Red", "Yellow", "Metal" };
+    public int weightInfluences(ICropTile crop, float humidity, float nutrients, float air) {
+        // no documentation so i'm leaving it alone instead of inverting the effect
+        return (int) Math.floor((double) humidity * 0.5 + (double) nutrients * 2.0 + (double) air * 0.5);
```
Comment (miozune) [inline src/main/java/com/github/bartimaeusnek/cropspp/crops/gregtechCrops/GarnydniaCrop.java:70]: Same here

File: src/main/java/com/github/bartimaeusnek/cropspp/crops/TC/BasicManaBeanCrop.java [hunk 8]
```diff
@@ -49,41 +65,34 @@ public int stat(int n) {
...
-            return crop.isBlockBelow(blockCrystal);
-        } else return true;
-    }
-
     @Override
     public int weightInfluences(ICropTile crop, float humidity, float nutrients, float air) {
         // Requires no humidity but nutrients.
-        return (int) ((double) humidity * 1.3 + (double) nutrients * 1 + (double) air * 0.7);
+        // Doc seems incorrect, but anyhow i've inverted it for safety sake
+        return (int) ((double) humidity / 1.3D + (double) nutrients + (double) air * 0.7);
```
Comment (miozune) [inline src/main/java/com/github/bartimaeusnek/cropspp/crops/TC/BasicManaBeanCrop.java:72]: Also make air requirement division maybe?

Comment (C0bra5) [reply src/main/java/com/github/bartimaeusnek/cropspp/crops/TC/BasicManaBeanCrop.java]: addressed in https://github.com/GTNewHorizons/Crops-plus-plus/pull/57/commits/363580f75b54ddb4688ce03d45a3a68421d83faa

File: src/main/java/com/github/bartimaeusnek/cropspp/crops/BoP/Bamboo.java [hunk 9]
```diff
@@ -11,23 +11,28 @@
...
     }
 
-    public int tier() {
-        return 2;
-    }
-
     @Override
     public int weightInfluences(ICropTile crop, float humidity, float nutrients, float air) {
+        // no documentation on the intended effect, so i'm leaving it as is
         return (int) ((double) humidity * 1.3D + (double) nutrients * 1.2D + (double) air * 0.8D);
```
[the remark this answers is not in the document]
Comment (C0bra5) [reply src/main/java/com/github/bartimaeusnek/cropspp/crops/BoP/Bamboo.java]: addressed in https://github.com/GTNewHorizons/Crops-plus-plus/pull/57/commits/363580f75b54ddb4688ce03d45a3a68421d83faa

File: src/main/java/com/github/bartimaeusnek/cropspp/crops/TC/PrimordialPearlBerryCrop.java [hunk 10]
```diff
@@ -45,19 +60,17 @@ public int stat(int n) {
...
-        return crop.getSize() < 4;
-    }
-
-    @Override
-    public boolean canBeHarvested(ICropTile crop) {
-        return crop.getSize() == 4;
+    public int weightInfluences(ICropTile crop, float humidity, float nutrients, float air) {
+        // Requires Tons of everything
+        // used to be able to grow with 2 humidity, 2 nutrients, and 2 air quality pretty well
+        // now this crop will be a propper pain in the ass to grow (about 4x slower
```
[the remark this answers is not in the document]
Comment (C0bra5) [reply src/main/java/com/github/bartimaeusnek/cropspp/crops/TC/PrimordialPearlBerryCrop.java]: addressed in https://github.com/GTNewHorizons/Crops-plus-plus/pull/57/commits/363580f75b54ddb4688ce03d45a3a68421d83faa

File: src/main/java/com/github/bartimaeusnek/cropspp/abstracts/BasicBerryCrop.java [hunk 11]
```diff
@@ -51,13 +46,14 @@ public List<String> getCropInformation() {
     @Override
     public int weightInfluences(ICropTile crop, float humidity, float nutrients, float air) {
         // Requires more humidity than nutrients or air, but not much more
-        return (int) ((double) humidity * 1.2 + (double) nutrients * 0.9 + (double) air * 0.9);
+        return (int) ((double) humidity / 1.2D + (double) nutrients / 0.9D + (double) air / 0.9D);
```
[the remark this answers is not in the document]
Comment (C0bra5) [reply src/main/java/com/github/bartimaeusnek/cropspp/abstracts/BasicBerryCrop.java]: addressed in https://github.com/GTNewHorizons/Crops-plus-plus/pull/57/commits/363580f75b54ddb4688ce03d45a3a68421d83faa

**Review by miozune — APPROVED** [review]
Sorry for the delayed review, great work!
Some commits could have been split into format changes and bug fixes, ~~but I cannot blame you because I also sometimes do that~~

**Pull request merged**